core/src/main/java/org/elasticsearch/index/query/BoolQueryParser.java

```diff
@@ -71,14 +71,12 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
                 switch (currentFieldName) {
                 case "must":
                     query = parseContext.parseInnerQueryBuilder();
-                    if (query != null) {
-                        mustClauses.add(query);
-                    }
+                    mustClauses.add(query);
                     break;
                 case "should":
                     query = parseContext.parseInnerQueryBuilder();
-                    if (query != null) {
-                        shouldClauses.add(query);
+                    shouldClauses.add(query);
+                    if (query != EmptyQueryBuilder.PROTOTYPE) {
                         if (parseContext.isFilter() && minimumShouldMatch == null) {
```

Comment:
If this BoolQuery is nested inside another queries "filter" section (e.g. another BoolQuerys "filter" part), then the parent query will call `parseInnerQuery()` which will in turn switch the `isFilter` flag. So I think we need it here as soon as we add a non-null should clause.